contrib/dhcp_agent/dhcp_agent.py

```diff
@@ -0,0 +1,350 @@
+        config_dir = os.path.expanduser('~/.ralph-dhcp-agent')
+        if not os.path.exists(config_dir):
+            os.makedirs(config_dir)
+        config_file = os.path.join(config_dir, 'config.db')
+        self.config = gdbm.open(config_file, 'c')
+
+    def save_config(self):
+        self.config.close()
+
+    def get_url_hash(self, url):
+        hash_url = hashlib.md5()
+        hash_url.update(url.encode())
+        return hash_url.hexdigest()
+
+    def get_last_modified(self, url):
+        h = self.get_url_hash(url)
+        try:
+            last = self.config[CONFIG_HTTP_LAST_MODIFIED + h]
+        except KeyError:
+            last = None
+        return last
+
+    def set_last_modified(self, url, value):
+        h = self.get_url_hash(url)
+        self.config[CONFIG_HTTP_LAST_MODIFIED + h] = value
+
+    def fetch_and_reload(self, mode, request_params):
+        dhcp_config = self._get_configuration(mode, request_params)
+        if dhcp_config:
+            self._set_new_configuration(dhcp_config, mode)
+            self._restart_dhcp_server()
+
+    def get_request_params(self, dcs, envs):
+        request_params = []
+        if envs:
+            request_params.extend([('env', env) for env in envs])
+        if dcs:
+            request_params.extend([('dc', dc) for dc in dcs])
+        assert request_params != []
+        return request_params
+
+    def _get_configuration(self, mode, params):
+        url = '{}://{}/dhcp/{}/?{}'.format(
+            self.proto,
+            self.host,
+            mode,
+            urlencode(params)
+        )
+        data = None
+
+        self.logger.info('Sending request to {}'.format(url))
+        try:
+            response = self.make_request(url)
+        except HTTPError as e:
+            if e.code != 304:
+                self.logger.error(
+                    'Server returned {} status code with message "{}"'.format(
+                        e.code, e.fp.read().decode()
+                    )
+                )
+                return False
+            else:
+                self.logger.info(
+                    'Server return status 304 NOT MODIFIED. Nothing to do.'
+                )
+                return
+        else:
+            data = response.read()
+            self.set_last_modified(url, response.headers.get('Last-Modified'))
+        return data
+
+    def _send_confirm(self):
+        url = '{}://{}/dhcp/synch/'.format(
+            self.proto,
+            self.host,
+        )
+        try:
+            self.make_request(url, data={})
+        except HTTPError as e:
+            self.logger.error(
+                'Could not send confirmation to Ralph. '
+                'Server returned {} status code with message: {}'.format(
+                    e.code, e.fp.read().decode()
+                ),
+            )
+            return False
+        self.logger.info('Confirmation sent to {}.'.format(self.host))
+        return True
```

Comment:
What's the meaning of the value returned by this function..? It is not immediately obvious.